### PR TITLE
[float8nocompile] Simplified Float8Linear implementation which only supports dynamic tensorwise scaling

### PR DESCRIPTION
**Pull request by danielvegamyhre**
Summary:

This PR adds a simplified implementation of Float8Linear dubbed `Float8LinearNoCompile` which only supports dynamic tensorwise scaling. I've used TODOs to mark the places where I need to replace the torch based logic with custom triton kernels, in order to improve eager mode performance. Once those kernels have been implemented, I'll benchmark the performance and do some profiling to identify bottlenecks and find additional optimization opportunities. 

The purpose of starting with  this is to start with a simple implementation which works e2e for float8 training (as shown in the test plan section below), so that as I start replacing torch logic with new triton kernels, I can use the e2e training example as a basic test to validate it's still working.

Test plan:
- Validated this simplified implementation works e2e by running a simple example in `examples/example.py`:

```
[danvm@devgpu006.vll6 /data/users/danvm/ao/torchao/prototype/float8nocompile/examples (linear1)]$ python3 example.py 
calling convert_to_float8_nocompile_training
finished convert_to_float8_nocompile_training
step 0
step 1
step 2
step 3
step 4
step 5
step 6
step 7
step 8
step 9
```

### CODE REVIEW

File: torchao/prototype/float8nocompile/examples/example.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+)
+x = torch.randn(32, 32, device="cuda", dtype=torch.bfloat16)
+optimizer = torch.optim.SGD(m.parameters(), lr=0.1)
+
+# convert specified `torch.nn.Linear` modules to `Float8Linear`
+print("calling convert_to_float8_nocompile_training")
+convert_to_float8_nocompile_training(m)
+print("finished convert_to_float8_nocompile_training")
+
+for i in range(10):
```
Comment (vkuzo) [inline torchao/prototype/float8nocompile/examples/example.py:28]: this makes sense, I'd recommend adding a stronger numerical equivalency check (can be in a future PR):
1. create example input
2. create (a) reference model and (b) your model
3. feed the example input, run backwards
4. compare that model output, weight gradients, input gradients are equivalent across (a) and (b)

File: torchao/prototype/float8nocompile/float8nocompile_linear.py [hunk 2]
```diff
@@ -19,4 +36,149 @@ class Float8LinearNoCompile(torch.nn.Linear):
...
 
-    pass
+    def __init__(self, *args, **kwargs):
+        """
+        Additional arguments on top of `torch.nn.Linear`'s arguments:
+        * `config`: Float8LinearConfig
+        """
+
+        # Amax scales should always be kept as float32.
+        self.always_float32_buffers = set()
```
Comment (vkuzo) [inline torchao/prototype/float8nocompile/float8nocompile_linear.py:46]: can delete, buffers are only needed for delayed scaling

File: torchao/prototype/float8nocompile/float8nocompile_linear.py [hunk 3]
```diff
@@ -19,4 +36,149 @@ class Float8LinearNoCompile(torch.nn.Linear):
...
+        config = kwargs.pop("config")
+        emulate = config.emulate
+        super().__init__(*args, **kwargs)
+
+        # Defines the scaling behavior of input, weight, grad_output
```
Comment (vkuzo) [inline torchao/prototype/float8nocompile/float8nocompile_linear.py:51]: can delete, this is only needed for supporting multiple types of scaling

File: torchao/prototype/float8nocompile/float8nocompile_linear.py [hunk 4]
```diff
@@ -19,4 +36,149 @@ class Float8LinearNoCompile(torch.nn.Linear):
...
+        self.scaling_type_input = config.cast_config_input.scaling_type
+        self.scaling_type_weight = config.cast_config_weight.scaling_type
+        self.scaling_type_grad_output = config.cast_config_grad_output.scaling_type
+
+        self.config = config
+        self.is_amax_initialized = not self.config.enable_amax_init
```
Comment (vkuzo) [inline torchao/prototype/float8nocompile/float8nocompile_linear.py:57]: can delete, this is only needed for delayed scaling

File: torchao/prototype/float8nocompile/float8nocompile_linear.py [hunk 5]
```diff
@@ -19,4 +36,149 @@ class Float8LinearNoCompile(torch.nn.Linear):
...
+                self.config.gemm_config_grad_weight.use_fast_accum,
+                False,
+                self.config.pad_inner_dim,
+            ),
+        )
+
+    def forward(self, input: torch.Tensor) -> torch.Tensor:
+        # TODO(danielvegamyhre): modify to support for FSDP once dependencies are implemented
+        output = self.forward_fp8_matmul(input)
+        if self.bias is not None:
```
Comment (vkuzo) [inline torchao/prototype/float8nocompile/float8nocompile_linear.py:86]: feel free to not support bias at all to simplify, as our target model uses linear without bias

File: torchao/prototype/float8nocompile/float8nocompile_linear.py [hunk 6]
```diff
@@ -19,4 +36,149 @@ class Float8LinearNoCompile(torch.nn.Linear):
...
+        return hp_tensor_to_float8nocompile_dynamic(
+            input,
+            self.config.cast_config_input.target_dtype,
+            self.linear_mm_config,
+            gemm_input_role=GemmInputRole.INPUT,
+        )
+
+    def get_weight_scale(self, weight: torch.Tensor) -> Optional[torch.Tensor]:
+        # TODO(danielvegamyhre): replace scale calculation with triton kernel
+        if tensor_already_casted_to_fp8(weight):
```
Comment (vkuzo) [inline torchao/prototype/float8nocompile/float8nocompile_linear.py:127]: feel free to remove this to simplify

File: torchao/prototype/float8nocompile/float8nocompile_linear.py [hunk 7]
```diff
@@ -19,4 +36,149 @@ class Float8LinearNoCompile(torch.nn.Linear):
...
+            output,
+            self.linear_mm_config,
+            self.config.cast_config_grad_output.target_dtype,
+        )
+
+    @classmethod
+    def from_float(
+        cls,
+        mod,
+        config: Optional[Float8LinearConfig] = None,
```
Comment (vkuzo) [inline torchao/prototype/float8nocompile/float8nocompile_linear.py:162]: nit: feel free to only support default settings and not give an option to modify the config

File: torchao/prototype/float8nocompile/float8nocompile_scaling_utils.py [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+        linear_mm_config: Defines the configuration for the scaled_mm for
+          the 3 fwd/bwd gemms of linear
+        reduce_amax: whether to reduce the max(abs(hp_tensor)) value across distributed ranks
+        gemm_input_role: Defines the role of this tensor (input, weight or grad_output) in
+          the 3 fwd/bwd gemms of linear
+        scaling_granularity: Defines the scaling granularity
+        axiswise_dim: if axiswise granularity is used, defines the dim to scale across
+    """
+    # TODO(danielvegamyhre): replace this torch implementation with custom triton kernel
+    if tensor_already_casted_to_fp8(hp_tensor):
```
Comment (vkuzo) [inline torchao/prototype/float8nocompile/float8nocompile_scaling_utils.py:52]: nit: can remove to simplify

File: torchao/prototype/float8nocompile/float8nocompile_scaling_utils.py [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+        return hp_tensor
+    scale = tensor_to_scale(
```
Comment (vkuzo) [inline torchao/prototype/float8nocompile/float8nocompile_scaling_utils.py:54]: I'd recommend just inlining the right code here for just the default settings instead of calling into util functions, it will be easier to compare that to handwritten kernels

File: torchao/prototype/float8nocompile/float8nocompile_scaling_utils.py [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+        hp_tensor,
+        float8_dtype,
+        reduce_amax,
+        device_mesh,
+        scaling_granularity,
+        axiswise_dim,
+    )
+    return hp_tensor_and_scale_to_float8(
```
Comment (vkuzo) [inline torchao/prototype/float8nocompile/float8nocompile_scaling_utils.py:62]: I'd recommend just inlining the right code here for just the default settings instead of calling into util functions, it will be easier to compare that to handwritten kernels

**Review by vkuzo — APPROVED** [review]
lgtm, feel free to address changes in this PR or future PRs

File: torchao/prototype/float8nocompile/examples/example.py (hunk 1 above)
Comment (danielvegamyhre) [reply torchao/prototype/float8nocompile/examples/example.py:28]: Makes sense, will do this in a follow up PR.

File: torchao/prototype/float8nocompile/float8nocompile_linear.py (hunk 2 above)
Comment (danielvegamyhre) [reply torchao/prototype/float8nocompile/float8nocompile_linear.py]: ah yes I noticed buffers weren't needed for dynamic scaling but missed this one somehow - thanks!

**Comment by danielvegamyhre** [comment]
> lgtm, feel free to address changes in this PR or future PRs

Sounds good, I addressed all your comments except the extension to the test script, which I'll do in a follow up PR - thanks!
